File: internal/proxy/fallback.go
package proxy

import (
	"bytes"
	"context"
	"encoding/json"
	rand "math/rand/v2"
	"net/http"
	"sync"
	"time"
)

// keyRing hands out API keys uniformly at random so concurrent keyed requests
// spread across the whole pool instead of following a fixed rotation order.
type keyRing struct {
	keys []string
}

func newKeyRing(keys []string) *keyRing {
	return &keyRing{keys: keys}
}

func (k *keyRing) pick() string {
	if k == nil || len(k.keys) == 0 {
		return ""
	}
	return k.keys[rand.IntN(len(k.keys))]
}

// modelFallback tracks the no-key -> API-key state for one upstream model.
type modelFallback struct {
	keyMode    bool
	noKeyFails int
	keyedFail  bool // keyed calls are currently failing
}

// state returns the effective status-page state of the model.
func (st *modelFallback) state() string {
	if st == nil || !st.keyMode {
		return stateAnonymous
	}
	if st.keyedFail {
		return stateKeyedFail
	}
	return stateKeyed
}

// fallbackState holds fallback state per upstream model: a 429 storm on one
// model (say model A) only switches that model to API-key mode, while other
// models keep going anonymously.
type fallbackState struct {
	mu     sync.RWMutex
	models map[string]*modelFallback
}

// keysEnabled reports whether the no-key -> API-key fallback is active.
func (p *Proxy) keysEnabled() bool {
	return p.keys != nil
}

// emitStateChange records a model's state transition and reports it to the
// status UI in the background (a no-op when no URL is configured).
func (p *Proxy) emitStateChange(model, from, to, reason, detail string) {
	if p.reporter == nil {
		return
	}
	p.reporter.send(StateEvent{
		Model:  model,
		From:   from,
		To:     to,
		Reason: reason,
		Detail: detail,
	})
}

// inKeyMode reports whether the given upstream model is currently in keyed mode.
func (p *Proxy) inKeyMode(model string) bool {
	if !p.keysEnabled() {
		return false
	}
	p.fb.mu.RLock()
	defer p.fb.mu.RUnlock()
	st := p.fb.models[model]
	return st != nil && st.keyMode
}

// keyedModels lists the upstream models currently in keyed mode.
func (p *Proxy) keyedModels() []string {
	if !p.keysEnabled() {
		return nil
	}
	p.fb.mu.RLock()
	defer p.fb.mu.RUnlock()
	var out []string
	for model, st := range p.fb.models {
		if st.keyMode {
			out = append(out, model)
		}
	}
	return out
}

// requestKey returns the API key to attach to the upstream call, if any.
// With the key file configured a key is used only when the model is in keyed
// mode, and one key is picked at random from the pool each time; otherwise the
// legacy single ZEN_UPSTREAM_API_KEY is used when set.
func (p *Proxy) requestKey(model string) (string, bool) {
	if !p.keysEnabled() {
		if p.cfg.UpstreamAPIKey != "" {
			return p.cfg.UpstreamAPIKey, true
		}
		return "", false
	}
	if p.inKeyMode(model) {
		return p.keys.pick(), true
	}
	return "", false
}

// recordNoKeyFailure counts one failed anonymous request for the model. It
// returns true and switches that model to keyed mode once the threshold is
// reached.
func (p *Proxy) recordNoKeyFailure(model string) (switched bool) {
	if !p.keysEnabled() {
		return false
	}
	p.fb.mu.Lock()
	defer p.fb.mu.Unlock()
	st := p.fb.models[model]
	if st == nil {
		st = &modelFallback{}
		p.fb.models[model] = st
	} else if st.keyMode {
		return false
	}
	st.noKeyFails++
	if st.noKeyFails >= p.cfg.NoKeyFailThreshold {
		p.log.Info("no-key upstream failing repeatedly, switching model to API-key mode",
			"model", model, "failures", st.noKeyFails)
		from := st.state()
		st.keyMode = true
		st.noKeyFails = 0
		p.emitStateChange(model, from, stateKeyed, "anonymous_failures",
			"anonymous requests failed repeatedly")
		return true
	}
	return false
}

// forceSwitchToKey immediately marks the model as keyed so the current
// request is transparently retried with an API key instead of leaking a
// non-2xx (e.g. 400/503) to the client. Returns false when no key pool is
// configured or the model is already keyed (a keyed retry that still errors
// is surfaced).
func (p *Proxy) forceSwitchToKey(model string) bool {
	if !p.keysEnabled() {
		return false
	}
	p.fb.mu.Lock()
	defer p.fb.mu.Unlock()
	st := p.fb.models[model]
	if st == nil {
		st = &modelFallback{}
		p.fb.models[model] = st
	}
	if st.keyMode {
		return false
	}
	p.log.Info("upstream non-2xx, switching model to API-key mode", "model", model)
	from := st.state()
	st.keyMode = true
	st.noKeyFails = 0
	p.emitStateChange(model, from, stateKeyed, "upstream_error",
		"non-2xx retries exhausted")
	return true
}

// recordNoKeySuccess resets the anonymous failure counter for the model.
func (p *Proxy) recordNoKeySuccess(model string) {
	if !p.keysEnabled() {
		return
	}
	p.fb.mu.Lock()
	defer p.fb.mu.Unlock()
	st := p.fb.models[model]
	if st == nil || st.keyMode {
		return
	}
	st.noKeyFails = 0
}

// trySwitchToNoKey flips the model back to anonymous mode; used by the
// recovery probe.
func (p *Proxy) trySwitchToNoKey(model string) bool {
	if !p.keysEnabled() {
		return false
	}
	p.fb.mu.Lock()
	defer p.fb.mu.Unlock()
	st := p.fb.models[model]
	if st == nil || !st.keyMode {
		return false
	}
	p.log.Info("no-key upstream recovered, switching model back to anonymous mode", "model", model)
	from := st.state()
	st.keyMode = false
	st.noKeyFails = 0
	st.keyedFail = false
	p.circuit.Reset()
	p.emitStateChange(model, from, stateAnonymous, "probe_recovered",
		"anonymous probe succeeded")
	return true
}

// reportKeyedResult tracks whether keyed calls for a model are currently
// failing and reports keyed_failed <-> keyed transitions to the status UI.
// Only changes are reported, so a sustained outage yields one down event and
// the following successful call one recovery event.
func (p *Proxy) reportKeyedResult(model string, ok bool, detail string) {
	if !p.keysEnabled() {
		return
	}
	p.fb.mu.Lock()
	st := p.fb.models[model]
	if st == nil || !st.keyMode {
		p.fb.mu.Unlock()
		return
	}
	from, to := "", ""
	switch {
	case ok && st.keyedFail:
		from, to = stateKeyedFail, stateKeyed
		st.keyedFail = false
	case !ok && !st.keyedFail:
		from, to = stateKeyed, stateKeyedFail
		st.keyedFail = true
	}
	p.fb.mu.Unlock()
	if to != "" {
		p.emitStateChange(model, from, to, "keyed_error", detail)
	}
}

// probeLoop periodically probes every keyed model anonymously (through the
// same transport, so via the socks5 proxy when configured) and switches each
// model back to anonymous mode as soon as the upstream accepts it.
func (p *Proxy) probeLoop() {
	t := time.NewTicker(p.cfg.NoKeyProbeInterval)
	defer t.Stop()
	for range t.C {
		for _, model := range p.keyedModels() {
			ctx, cancel := context.WithTimeout(context.Background(), p.cfg.NoKeyProbeInterval)
			ok := p.probeAnonymous(ctx, model)
			cancel()
			if ok {
				p.trySwitchToNoKey(model)
			}
		}
	}
}

// probeAnonymous sends a tiny chat completion for the given model without any
// API key and reports whether the upstream accepted it.
func (p *Proxy) probeAnonymous(ctx context.Context, model string) bool {
	if model == "" {
		if len(p.cfg.Models) > 0 {
			model = p.cfg.Models[0]
		} else {
			model = "deepseek-v4-flash-free"
		}
	}
	body, _ := json.Marshal(map[string]any{
		"model":      model,
		"messages":   []map[string]string{{"role": "user", "content": "ping"}},
		"max_tokens": 1,
	})
	req, err := http.NewRequestWithContext(ctx, http.MethodPost,
		p.cfg.UpstreamBase+"/chat/completions", bytes.NewReader(body))
	if err != nil {
		return false
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "zen-proxy/1.0")
	resp, err := p.client.Do(req)
	if err != nil {
		p.log.Debug("no-key probe failed", "model", model, "error", err)
		return false
	}
	defer resp.Body.Close()
	return resp.StatusCode >= 200 && resp.StatusCode < 300
}

File: src/proxy/fallback.ts
/**
 * Per-model no-key -> API-key fallback. A 429 storm on one model only
 * switches that model to keyed mode; other models keep going anonymously.
 *
 * Recovery follows a hold-then-confirm flow: after a confirmed failure the
 * model stays in keyed mode for a hold window (default 5 minutes) with no
 * anonymous probes at all; once the hold expires the probe loop runs every
 * ZEN_NO_KEY_PROBE_SECONDS and only switches back to anonymous after
 * ZEN_NO_KEY_RECOVERY_CONFIRMATIONS consecutive successful probes, so a
 * flaky upstream cannot flap the model back and forth.
 */
import type { Logger } from "../log";
import type { Reporter } from "./reporter";

export type ModelState = "anonymous" | "keyed" | "keyed_failed";

interface ModelFallback {
  keyMode: boolean;
  noKeyFails: number;
  keyedFail: boolean;
  keyedAt: number;
  probeStreak: number;
}

export class FallbackState {
  private models = new Map<string, ModelFallback>();
  /** Model -> timestamp of its most recent route through this proxy, so the
   *  status UI can surface exactly the models that were actually called
   *  within the 24h window. */
  private seen = new Map<string, number>();
  constructor(
    private threshold: number,
    private keysEnabled: boolean,
    private log: Logger,
    private reporter?: Reporter,
    private holdMs: number = 5 * 60 * 1000,
    private confirmations: number = 3,
    private nowFn: () => number = () => Date.now(),
  ) {}

  state(model: string): ModelState {
    const st = this.models.get(model);
    if (!st || !st.keyMode) return "anonymous";
    return st.keyedFail ? "keyed_failed" : "keyed";
  }

  inKeyMode(model: string): boolean {
    if (!this.keysEnabled) return false;
    return this.models.get(model)?.keyMode === true;
  }

  keyedModels(): string[] {
    if (!this.keysEnabled) return [];
    const out: string[] = [];
    for (const [model, st] of this.models) {
      if (st.keyMode) out.push(model);
    }
    return out;
  }

  knownModels(): string[] {
    return [...this.seen.keys()];
  }

  /** Unix ms of the most recent time this model was routed (0 if never). */
  lastSeenAt(model: string): number {
    return this.seen.get(model) ?? 0;
  }

  /** Record that this model was routed right now. */
  private touch(model: string) {
    this.seen.set(model, this.nowFn());
  }

  /** Current clock value used for routing timestamps (injectable in tests). */
  clock(): number {
    return this.nowFn();
  }

  /** Public entry point used by the upstream client / handler to mark that
   *  a request for this model reached the routing path (before any retry or
   *  fallback decision), so "called" means "attempted", not only "succeeded". */
  noteCall(model: string) {
    this.touch(model);
  }

  recordNoKeyFailure(model: string): boolean {
    if (!this.keysEnabled) return false;
    this.touch(model);
    let st = this.models.get(model);
    if (!st) {
      st = { keyMode: false, noKeyFails: 0, keyedFail: false, keyedAt: 0, probeStreak: 0 };
      this.models.set(model, st);
    } else if (st.keyMode) {
      return false;
    }
    st.noKeyFails++;
    if (st.noKeyFails >= this.threshold) {
      this.log.info("no-key upstream failing repeatedly, switching model to API-key mode", { model, failures: st.noKeyFails });
      st.keyMode = true;
      st.noKeyFails = 0;
      st.keyedAt = this.nowFn();
      st.probeStreak = 0;
      this.emit(model, "anonymous", "keyed", "anonymous_failures", "anonymous requests failed repeatedly");
      return true;
    }
    return false;
  }

  forceSwitchToKey(model: string): boolean {
    if (!this.keysEnabled) return false;
    this.touch(model);
    let st = this.models.get(model);
    if (!st) {
      st = { keyMode: false, noKeyFails: 0, keyedFail: false, keyedAt: 0, probeStreak: 0 };
      this.models.set(model, st);
    }
    if (st.keyMode) return false;
    this.log.info("upstream non-2xx, switching model to API-key mode", { model });
    st.keyMode = true;
    st.noKeyFails = 0;
    st.keyedAt = this.nowFn();
    st.probeStreak = 0;
    this.emit(model, "anonymous", "keyed", "upstream_error", "first non-2xx response");
    return true;
  }

  recordNoKeySuccess(model: string) {
    this.touch(model);
    if (!this.keysEnabled) return;
    let st = this.models.get(model);
    if (!st) {
      this.models.set(model, { keyMode: false, noKeyFails: 0, keyedFail: false, keyedAt: 0, probeStreak: 0 });
      return;
    }
    if (st.keyMode) return;
    st.noKeyFails = 0;
  }

  takeProbeResult(model: string, ok: boolean): boolean {
    this.touch(model);
    if (!this.keysEnabled) return false;
    if (!this.probesActive(model)) return false;
    const st = this.models.get(model);
    if (!st || !st.keyMode) return false;
    if (ok) {
      // A keyed success does not reset the streak: intermediate trafficless
      // gaps and held key-mode periods must not undo accumulated probe
      // successes. Only a failed probe (or leaving key mode) resets it.
      const next = st.probeStreak + 1;
      const done = next >= this.confirmations;
      this.log.debug("no-key probe succeeded", { model, streak: next, confirmations: this.confirmations });
      if (done) {
        this.log.info("no-key upstream recovered, switching model back to anonymous mode", { model, confirmations: next });
        st.keyMode = false;
        st.noKeyFails = 0;
        st.keyedFail = false;
        st.keyedAt = 0;
        st.probeStreak = 0;
        this.emit(model, "keyed", "anonymous", "probe_recovered", "anonymous probe succeeded");
        return true;
      }
      st.probeStreak = next;
      return false;
    }
    if (st.probeStreak > 0) {
      this.log.info("no-key probe failed, restarting recovery streak", { model, streak: st.probeStreak });
    }
    st.probeStreak = 0;
    return false;
  }

  /** Whether the probe loop should ping this model right now: it must be in
   *  key mode, the hold window must have elapsed, and it must still be short
   *  of the confirmation streak. While the hold window is active no anonymous
   *  probes run at all; once it expires probing resumes every
   *  ZEN_NO_KEY_PROBE_SECONDS until N consecutive probes succeed. */
  probesActive(model: string): boolean {
    if (!this.keysEnabled) return false;
    const st = this.models.get(model);
    if (!st || !st.keyMode) return false;
    if (this.nowFn() - st.keyedAt < this.holdMs) return false; // inside the hold window
    return st.probeStreak < this.confirmations;
  }

  reportKeyedResult(model: string, ok: boolean, detail: string) {
    if (!this.keysEnabled) return;
    this.touch(model);
    const st = this.models.get(model);
    if (!st || !st.keyMode) return;
    let from = "";
    let to = "";
    if (ok && st.keyedFail) {
      from = "keyed_failed";
      to = "keyed";
      st.keyedFail = false;
    } else if (!ok && !st.keyedFail) {
      from = "keyed";
      to = "keyed_failed";
      st.keyedFail = true;
    }
    if (to) this.emit(model, from, to, "keyed_error", detail);
  }

  private emit(model: string, from: string, to: string, reason: string, detail: string) {
    this.reporter?.send({ model, from, to, reason, detail });
  }
}
